### DISCUSSION TITLE
App Store Connect API - using JWT

### DISCUSSION THREAD

**Discussion by AlesMMichalek** [fastlane/fastlane#18614] [Q&A] [answered]
Hello.

   We are planning to use App Store Connect API with API key to distribute our application to App Store. As we know, the fastlane allows this (https://docs.fastlane.tools/app-store-connect-api/) by calling method app_store_connect_api_key. Unfortunately, it requires direct API key (private-key) exposure to fastlane, which is against our security policy.
By Apple documentation we are able to pre-prepare necessary API authentication token (JWT) securely. Is there some way, how to pass the generated token to fastlane for necessary services? For example this way:

```
lane :release do
   app_store_connect_api_token(path: 'path/to/jwt')
   pilot
end
```

or 

```
lane :relase do
   app_store_connect_api_token(token: 'eyJhbGciOiJIUzI1NiIsIndjakdj.askldjalsdjlajsdla.aklsdjakldsjlaj')
   pilot
end
```

If it is possible, could you please give me some advice how to do it? Or if it is not, do you think it could be added as new feature? Thanks a lot.

**Comment by max-ott** [comment, marked as answer]
It's not possible because the token expires after 20 minutes. `Fastlane` takes care of generating the token and making sure it's valid.

> The token's expiration time, in Unix epoch time; tokens that expire more than 20 minutes in the future are not valid
(Source: https://developer.apple.com/documentation/appstoreconnectapi/generating_tokens_for_api_requests)

There are multiple options of storing it in environment variables and similar for security purposes, e.g. https://docs.fastlane.tools/app-store-connect-api/#using-fastlane-api-key-json-file -> The content of the file as ENV var or secret or similar.

**Reply by AlesMMichalek** [reply]
Thanks for your answer.

I think 20 minutes is far enough to perform required action(s). Generating fresh JWT is responsibility of command caller. I do not see any issue here: If automation needs to perform fastlane action, internal service will be asked for fresh JWT and it will pass token to fastlane to perform action (in fact there could be special fastlane action to request JWT from some remote service). Without any offence, it is always bad habit to pass secret (especially private key) to other subjects, e.g. 3rd party tool, regardless of way (plain file, script, env variable). I think this is true purpose why Apple introduced JWT - to have separated (less secure) tooling for deployment and very secure private-key handling. Token short expiration period extremely limits misusing of token in case of compromising. 

To be honest, I am not able to make code change myself, but I guess it is not big deal for experienced fastlane contributors. Do you really think such feature will not be welcome? Support of externally generated token?

**Reply by max-ott** [reply]
I‘m not seeing any value in adding this. This would require changes to the login flow for an extreme edge case. Feel free to open a pull request or pay a freelancer to do it.

With fastlane all information stays on the local machine. If you are concerned about security, simply don’t use cloud based CI/CD and run everything on your own managed servers. There you have full control over everything. Don’t expect anyone from the core team to work on this.
